src/ODESolvers/MultirateInfinitesimalStepMethod.jl

```diff
@@ -0,0 +1,276 @@
+  "fast rhs method"
+  fastsolver::FS
+  "number of substeps per stage"
+  nsubsteps::Int
+  α::SArray{NTuple{2, Nstages}, RT, 2, Nstages_sq} 
+  β::SArray{NTuple{2, Nstages}, RT, 2, Nstages_sq}
+  γ::SArray{NTuple{2, Nstages}, RT, 2, Nstages_sq}
+  d::SArray{NTuple{1, Nstages}, RT, 1, Nstages} 
+  c::SArray{NTuple{1, Nstages}, RT, 1, Nstages} 
+  c̃::SArray{NTuple{1, Nstages}, RT, 1, Nstages} 
+  function MultirateInfinitesimalStep(slowrhs!, fastrhs!, fastmethod, nsubsteps,
+                                      α, β, γ,
+                                      Q::AT; dt=0, t0=0) where {AT<:AbstractArray}
+
+    T = eltype(Q)
+    RT = real(T)
+    dt = [dt]
+    t0 = [t0]
+
+    Nstages = size(α, 1)
+
+    yn = similar(Q)
+    ΔYnj = ntuple(_ -> similar(Q), Nstages-2)
+    fYnj = ntuple(_ -> similar(Q), Nstages-1)
+    offset = similar(Q)
+    tsfastrhs! = TimeScaledRHS(RT(0), RT(0), fastrhs!)
+    fastsolver = fastmethod(tsfastrhs!, Q)
+
+    d = sum(β, dims=2)
+
+    c = similar(d)
+    for i = eachindex(c)
+      c[i] = d[i]
+      if i > 1
+        c[i] += sum(j-> (α[i,j] + γ[i,j])*c[j], 1:i-1)
+      end
+    end
+    c̃ = α*c
+    
+    new{T, RT, AT, typeof(fastsolver), Nstages, Nstages-1, Nstages-2, Nstages ^ 2}(dt, t0, yn, ΔYnj, fYnj, offset,
+                                         slowrhs!, tsfastrhs!, fastsolver, nsubsteps,
+                                         α, β, γ, d, c, c̃)
+  end
+end
+
+
+# TODO almost identical functions seem to be defined for every ode solver,
+# define a common one in ODEsolvers ?
+function ODEs.dostep!(Q, mis::MultirateInfinitesimalStep, param,
+                      timeend::AbstractFloat, adjustfinalstep::Bool)
+  time, dt = mis.t[1], mis.dt[1]
+  @assert dt > 0
+  if adjustfinalstep && time + dt > timeend
+    dt = timeend - time
+    @assert dt > 0
+  end
+
+  ODEs.dostep!(Q, mis, param, time, dt)
+
+  if dt == mis.dt[1]
+    mis.t[1] += dt
+  else
+    mis.t[1] = timeend
+  end
+  return mis.t[1]
+end
+
+function ODEs.dostep!(Q, mis::MultirateInfinitesimalStep, p,
+                      time::AbstractFloat, dt::AbstractFloat)
+  FT = eltype(dt)
+  α = mis.α
+  β = mis.β
+  γ = mis.γ
+  yn = mis.yn
+  ΔYnj = mis.ΔYnj
+  fYnj = mis.fYnj
+  offset = mis.offset
+  d = mis.d
+  c = mis.c
+  c̃ = mis.c̃
+  slowrhs! = mis.slowrhs!
+  fastsolver = mis.fastsolver
+  fastrhs! = mis.tsfastrhs!
+
+  nstages = size(α, 1)
+
+  copyto!(yn, Q) # first stage
+  for i = 2:nstages
+    slowrhs!(fYnj[i-1], Q, p, time + c[i-1]*dt, increment=false)
+
+    threads = 256
+    blocks = div(length(realview(Q)) + threads - 1, threads)
+    @launch(device(Q), threads=threads, blocks=blocks,
+            update!(realview(Q), realview(offset), Val(i), realview(yn), map(realview, ΔYnj[1:i-2]), map(realview, fYnj[1:i-1]), 
+            α[i,:], β[i,:], γ[i,:], d[i], dt))
+
+    fastrhs!.a = time + c̃[i]*dt
+    fastrhs!.b = (c[i] - c̃[i]) / d[i]
+
+    τ = zero(FT)
+    dτ = d[i] * dt / mis.nsubsteps
```

Comment:
Should number of substeps be hard coded, or should a fast dt be specified and the number of substeps be chosen to match that fast dt?

Reply:
My understanding is that we will typically specify everything in terms of ratio of steps, and doing it this way avoids the need to worry about floating point rounding, but we could probably support both?

Reply:
My thought here is that the number of substeps will be dictated by the largest value in the `d` array. If they are not all about the same size, you could end up substepping the stages with smaller `d_i` values using many more steps that needed for stability/accuracy.

For the LSRK, the ratio was about `~75` for the 14 stage LSRK so we'd do substantially more work for the small partial steps than we needed to.

A quick check suggested the situation is not as bad for the methods you all implemented here.

You can see how I handled it [here](https://github.com/climate-machine/CLIMA/blob/72c35b7687601ad0669861ef36b199223851e3f8/src/ODESolvers/MultirateRungeKuttaMethod.jl#L153-L154)

Reply:
Also, realized that the ratio of steps is the outer to inner full step size, not the step size needed to substep the stage values.